File: core/src/test/java/com/kin/ecosystem/core/data/offer/OfferRepositoryTest.java
package com.kin.ecosystem.core.data.offer;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import com.kin.ecosystem.common.KinCallback;
import com.kin.ecosystem.common.NativeOfferClickEvent;
import com.kin.ecosystem.common.Observer;
import com.kin.ecosystem.common.Callback;
import com.kin.ecosystem.common.model.NativeEarnOffer;
import com.kin.ecosystem.common.model.NativeEarnOfferBuilder;
import com.kin.ecosystem.common.model.NativeOffer;
import com.kin.ecosystem.common.model.NativeOffer.OfferType;
import com.kin.ecosystem.common.model.NativeOfferBuilder;
import com.kin.ecosystem.common.model.NativeSpendOffer;
import com.kin.ecosystem.common.model.NativeSpendOfferBuilder;
import com.kin.ecosystem.core.data.order.OrderDataSource;
import com.kin.ecosystem.common.exception.KinEcosystemException;
import com.kin.ecosystem.core.network.model.Offer;
import com.kin.ecosystem.core.network.model.OfferList;
import com.kin.ecosystem.core.network.model.Paging;
import com.kin.ecosystem.core.network.model.PagingCursors;
import com.kin.ecosystem.core.util.OfferConverter;
import java.lang.reflect.Field;
import java.util.ArrayList;
import java.util.List;
import com.kin.ecosystem.core.network.ApiException;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentCaptor;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.annotation.Config;

@Config(manifest = Config.NONE)
@RunWith(RobolectricTestRunner.class)
public class OfferRepositoryTest {

	@Mock
	private OrderDataSource orderRepository;

	@Mock
	private OfferDataSource.Remote remote;

	@Mock
	private Offer offer;

	private OfferRepository offerRepository;

	@Before
	public void setUp() throws Exception {
		MockitoAnnotations.initMocks(this);
		resetInstance();
	}

	private void resetInstance() throws Exception {
		Field instance = OfferRepository.class.getDeclaredField("instance");
		instance.setAccessible(true);
		instance.set(null, null);
		OfferRepository.init(remote, orderRepository);
		offerRepository = OfferRepository.getInstance();

		when(offer.getId()).thenReturn("1");
		when(offer.getAmount()).thenReturn(10);
	}

	@Test
	public void getCachedOfferList_Empty() {
		List<Offer> offers = offerRepository.getCachedOfferList().getOffers();
		assertEquals(0, offers.size());
	}

	@Test
	public void getOffers_Succeed_SavedToCachedList() {
		KinCallback<OfferList> offerListCallback = mock(KinCallback.class);
		ArgumentCaptor<Callback<OfferList, ApiException>> getOfferCapture = ArgumentCaptor.forClass(Callback.class);

		OfferList offerList = getOfferList();

		offerRepository.getOffers(offerListCallback);
		verify(remote).getOffers(getOfferCapture.capture());
		getOfferCapture.getValue().onResponse(offerList);

		assertEquals(1, offerRepository.getCachedOfferList().getOffers().size());
		verify(offerListCallback).onResponse(offerList);
	}

	@Test
	public void getOffers_Failed() {
		KinCallback<OfferList> offerListCallback = mock(KinCallback.class);
		ArgumentCaptor<Callback<OfferList, ApiException>> getOfferCapture = ArgumentCaptor.forClass(Callback.class);

		offerRepository.getOffers(offerListCallback);
		verify(remote).getOffers(getOfferCapture.capture());

		getOfferCapture.getValue().onFailure(getApiException());
		assertEquals(0, offerRepository.getCachedOfferList().getOffers().size());
		verify(offerListCallback).onFailure(any(KinEcosystemException.class));
	}

	@Test
	public void addNativeOfferCallback() throws Exception {
		Observer<NativeOfferClickEvent> callback = new Observer<NativeOfferClickEvent>() {
			@Override
			public void onChanged(NativeOfferClickEvent nativeOffer) {
				assertEquals("5", nativeOffer.getNativeOffer().getId());
				assertFalse(nativeOffer.isDismissOnTap());
			}
		};

		offerRepository.addNativeOfferClickedObserver(callback);
		offerRepository.getNativeSpendOfferObservable().postValue(new NativeOfferClickEvent.Builder()
			.nativeOffer(new NativeEarnOffer("5"))
			.isDismissed(false)
			.build());
	}

	@Test
	public void addNativeSpendOffer() throws Exception {

		List<Offer> newEarnOffers = new ArrayList<>();
		List<Offer> newSpendOffers = new ArrayList<>();
		NativeOffer nativeOffer =
			new NativeSpendOfferBuilder("1")
				.title("Native offer title")
				.description("Native offer desc")
				.amount(1000)
				.image("Native offer image").build();

		offerRepository.addNativeOffer(nativeOffer, true);
		assertEquals(1, offerRepository.getCachedOfferList().getOffers().size());
		assertEquals(nativeOffer.getId(), offerRepository.getCachedOfferList().getOffers().get(0).getId());

		Offer offer = OfferConverter.toOffer(nativeOffer);
		assertTrue(offerRepository.shouldDismissOnTap(offer.getId()));

		// Update on second time, still the size is 1 with same offer
		offerRepository.addNativeOffer(nativeOffer, false);
		offer = OfferConverter.toOffer(nativeOffer);
		assertFalse(offerRepository.shouldDismissOnTap(offer.getId()));
		assertEquals(1, offerRepository.getCachedOfferList().getOffers().size());
		assertEquals(nativeOffer.getId(), offerRepository.getCachedOfferList().getOffers().get(0).getId());
		OfferListUtil.splitOffersByType(offerRepository.getCachedOfferList().getOffers(), newEarnOffers, newSpendOffers);
		assertEquals(1, newSpendOffers.size());
		assertEquals(0, newEarnOffers.size());
	}


	@Test
	public void addNativeEarnOffer() throws Exception {

		List<Offer> newEarnOffers = new ArrayList<>();
		List<Offer> newSpendOffers = new ArrayList<>();
		NativeOffer nativeOffer =
			new NativeEarnOfferBuilder("2")
				.title("Native offer title")
				.description("Native offer desc")
				.amount(1000)
				.image("Native offer image").build();

		offerRepository.addNativeOffer(nativeOffer, true);
		assertEquals(1, offerRepository.getCachedOfferList().getOffers().size());
		assertEquals(nativeOffer.getId(), offerRepository.getCachedOfferList().getOffers().get(0).getId());

		Offer offer = OfferConverter.toOffer(nativeOffer);
		assertTrue(offerRepository.shouldDismissOnTap(offer.getId()));

		// Update on second time, still the size is 1 with same offer
		offerRepository.addNativeOffer(nativeOffer, false);
		offer = OfferConverter.toOffer(nativeOffer);
		assertFalse(offerRepository.shouldDismissOnTap(offer.getId()));
		assertEquals(1, offerRepository.getCachedOfferList().getOffers().size());
		assertEquals(nativeOffer.getId(), offerRepository.getCachedOfferList().getOffers().get(0).getId());
		OfferListUtil.splitOffersByType(offerRepository.getCachedOfferList().getOffers(), newEarnOffers, newSpendOffers);
		assertEquals(0, newSpendOffers.size());
		assertEquals(1, newEarnOffers.size());
	}



	@Test
	public void removeNativeSpendOffer() throws Exception {
		NativeOffer nativeOffer =
			new NativeSpendOfferBuilder("1")
				.title("Native offer title")
				.description("Native offer desc")
				.amount(1000)
				.image("Native offer image")
				.build();

		offerRepository.addNativeOffer(nativeOffer, true);
		assertEquals(1, offerRepository.getCachedOfferList().getOffers().size());
		assertEquals(nativeOffer.getId(), offerRepository.getCachedOfferList().getOffers().get(0).getId());

		Offer offer = OfferConverter.toOffer(nativeOffer);
		assertTrue(offerRepository.shouldDismissOnTap(offer.getId()));

		offerRepository.removeNativeOffer(nativeOffer);
		assertEquals(0, offerRepository.getCachedOfferList().getOffers().size());
	}

	@Test
	public void shouldDismissOnTap_WithNoExistingOfferId() {
		assertFalse(offerRepository.shouldDismissOnTap("404"));
	}

	private OfferList getOfferList() {
		OfferList offerList = new OfferList();
		offerList.add(offer);
		offerList.setPaging(new Paging().next("1").previous("0").cursors(new PagingCursors().after("1").before("0")));
		return offerList;
	}

	private ApiException getApiException() {
		Exception exception = new IllegalArgumentException();
		ApiException apiException = new ApiException(500, exception);
		return apiException;
	}
}

File: core/src/test/kotlin/com/kin/ecosystem/core/data/offer/OfferRepositoryTest.kt
package com.kin.ecosystem.core.data.offer

import com.kin.ecosystem.common.Callback
import com.kin.ecosystem.common.KinCallback
import com.kin.ecosystem.common.NativeOfferClickEvent
import com.kin.ecosystem.common.Observer
import com.kin.ecosystem.common.exception.KinEcosystemException
import com.kin.ecosystem.common.model.NativeEarnOffer
import com.kin.ecosystem.common.model.NativeEarnOfferBuilder
import com.kin.ecosystem.common.model.NativeSpendOfferBuilder
import com.kin.ecosystem.core.data.order.OrderDataSource
import com.kin.ecosystem.core.network.ApiException
import com.kin.ecosystem.core.network.model.Offer
import com.kin.ecosystem.core.network.model.OfferList
import com.kin.ecosystem.core.network.model.Paging
import com.kin.ecosystem.core.network.model.PagingCursors
import com.kin.ecosystem.core.util.OfferConverter
import com.nhaarman.mockitokotlin2.argumentCaptor
import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.whenever
import org.junit.Assert.assertEquals
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.ArgumentMatchers.any
import org.mockito.Mockito.verify
import org.mockito.MockitoAnnotations
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config

@Config(manifest = Config.NONE)
@RunWith(RobolectricTestRunner::class)
class OfferRepositoryTest {

    private val orderRepository: OrderDataSource = mock()
    private val remote: OfferDataSource.Remote = mock()
    private val offer: Offer = mock()

    private lateinit var offerRepository: OfferRepository

    private val offerList: OfferList
        get() {
            val offerList = OfferList()
            offerList.add(offer)
            offerList.paging = Paging().next("1").previous("0").cursors(PagingCursors().after("1").before("0"))
            return offerList
        }

    private val apiException: ApiException
        get() {
            val exception = IllegalArgumentException()
            return ApiException(500, exception)
        }

    @Before
    @Throws(Exception::class)
    fun setUp() {
        MockitoAnnotations.initMocks(this)
        resetInstance()
    }

    @Throws(Exception::class)
    private fun resetInstance() {
        val instance = OfferRepository::class.java.getDeclaredField("instance")
        instance.isAccessible = true
        instance.set(null, null)
        OfferRepository.init(remote, orderRepository)
        offerRepository = OfferRepository.getInstance()

        with(offer) {
            whenever(id).thenReturn("1")
            whenever(amount).thenReturn(10)
        }
    }

    @Test
    fun `first time cached offer list is empty`() {
        val offers = offerRepository.cachedOfferList.offers
        assertEquals(0, offers.size)
    }

    @Test
    fun `get offers from remote and update cached offers on success`() {
        val offerListCallback : KinCallback<OfferList> = mock()
        val getOfferCapture = argumentCaptor<Callback<OfferList, ApiException>>()

        val offerList = offerList

        offerRepository.getOffers(offerListCallback)
        verify(remote).getOffers(getOfferCapture.capture())
        getOfferCapture.firstValue.onResponse(offerList)

        assertEquals(1, offerRepository.cachedOfferList.offers.size)
        verify<KinCallback<OfferList>>(offerListCallback).onResponse(offerList)
    }

    @Test
    fun `get offers from remote failde, cached list is empty`() {
        val offerListCallback : KinCallback<OfferList> = mock()
        val getOfferCapture = argumentCaptor<Callback<OfferList, ApiException>>()

        offerRepository.getOffers(offerListCallback)
        verify(remote).getOffers(getOfferCapture.capture())

        getOfferCapture.firstValue.onFailure(apiException)
        assertEquals(0, offerRepository.cachedOfferList.offers.size)
        verify(offerListCallback).onFailure(any(KinEcosystemException::class.java))
    }

    @Test
    fun `add native offer callback and listen to post values`() {
        val callback: Observer<NativeOfferClickEvent> = mock()
        val offer = NativeOfferClickEvent.Builder()
                .nativeOffer(NativeEarnOffer("5"))
                .isDismissed(false)
                .build()

        offerRepository.addNativeOfferClickedObserver(callback)
        offerRepository.nativeSpendOfferObservable.postValue(offer)

        verify(callback).onChanged(offer)
    }

    @Test
    fun `add native spend offer to list`() {
        val nativeOffer = NativeSpendOfferBuilder("1")
                .title("Native offer title")
                .description("Native offer desc")
                .amount(1000)
                .dismissOnTap(true)
                .image("Native offer image").build()

        offerRepository.addNativeOffer(nativeOffer)
        with(offerRepository.cachedOfferList.offers) {
            assertEquals(1, size.toLong())
            assertEquals(OfferConverter.toOffer(nativeOffer), this[0])
        }
    }

    @Test
    fun `add native earn offer to list`() {
        val nativeOffer = NativeEarnOfferBuilder("2")
                .title("Native offer title")
                .description("Native offer desc")
                .amount(1000)
                .dismissOnTap(true)
                .image("Native offer image").build()

        offerRepository.addNativeOffer(nativeOffer)
        with(offerRepository.cachedOfferList.offers) {
            assertEquals(1, size.toLong())
            assertEquals(OfferConverter.toOffer(nativeOffer), this[0])
        }
    }

    @Test
    fun `remove native offer from list`() {
        val nativeOffer = NativeSpendOfferBuilder("1")
                .title("Native offer title")
                .description("Native offer desc")
                .amount(1000)
                .image("Native offer image")
                .dismissOnTap(true)
                .build()

        offerRepository.addNativeOffer(nativeOffer)
        with(offerRepository.cachedOfferList.offers) {
            assertEquals(1, size.toLong())
            assertEquals(OfferConverter.toOffer(nativeOffer), this[0])
        }


        offerRepository.removeNativeOffer(nativeOffer)
        assertEquals(0, offerRepository.cachedOfferList.offers.size)
    }

    @Test
    fun `add all offers to the top of the list and keep the order`() {
        val firstOrder = NativeSpendOfferBuilder("1")
                .title("First Native offer title")
                .description("Native offer desc")
                .amount(1)
                .image("Native offer image")
                .dismissOnTap(true)
                .build()

        val secondOrder = NativeSpendOfferBuilder("2")
                .title("Second Native offer title")
                .description("Native offer desc")
                .amount(2)
                .image("Native offer image")
                .dismissOnTap(false)
                .build()

        val thirdOrder = NativeSpendOfferBuilder("3")
                .title("Third Native offer title")
                .description("Native offer desc")
                .amount(3)
                .image("Native offer image")
                .dismissOnTap(true)
                .build()

        offerRepository.addNativeOffer(thirdOrder)
        offerRepository.addAllNativeOffers(arrayListOf(firstOrder, secondOrder))
        with(offerRepository.cachedOfferList.offers) {
            assertEquals(3, size.toLong())
            assertEquals(OfferConverter.toOffer(firstOrder), this[0])
            assertEquals(OfferConverter.toOffer(secondOrder), this[1])
            assertEquals(OfferConverter.toOffer(thirdOrder), this[2])
        }
    }

    @Test
    fun `update native offer if exists`() {
        val nativeOffer = NativeEarnOfferBuilder("2")
                .title("Native offer title")
                .description("Native offer desc")
                .amount(1000)
                .dismissOnTap(true)
                .image("Native offer image").build()

        offerRepository.addNativeOffer(nativeOffer)
        with(offerRepository.cachedOfferList.offers) {
            assertEquals(1, size.toLong())
            assertEquals(OfferConverter.toOffer(nativeOffer), this[0])
        }

        nativeOffer.amount = 5
        nativeOffer.isDismissOnTap = false
        offerRepository.addNativeOffer(nativeOffer)
        with(offerRepository.cachedOfferList.offers) {
            assertEquals(1, size.toLong())
            assertEquals(OfferConverter.toOffer(nativeOffer), this[0])
        }

        nativeOffer.amount = 22
        nativeOffer.isDismissOnTap = true
        offerRepository.addAllNativeOffers(arrayListOf(nativeOffer))
        with(offerRepository.cachedOfferList.offers) {
            assertEquals(1, size.toLong())
            assertEquals(OfferConverter.toOffer(nativeOffer), this[0])
        }
    }

    @Test
    fun `always add offer to top of the list`() {
        val firstOffer = NativeEarnOfferBuilder("1")
                .title("First Native offer title")
                .description("Native offer desc")
                .amount(1)
                .dismissOnTap(true)
                .image("Native offer image").build()

        val secondOffer = NativeEarnOfferBuilder("2")
                .title("Second Native offer title")
                .description("Native offer desc")
                .amount(2)
                .dismissOnTap(false)
                .image("Native offer image").build()

        offerRepository.addNativeOffer(secondOffer)
        offerRepository.addNativeOffer(firstOffer)

        //The last order added should be on the top of the list
        with(offerRepository.cachedOfferList.offers) {
            assertEquals(2, size.toLong())
            assertEquals(OfferConverter.toOffer(firstOffer), this[0])
            assertEquals(OfferConverter.toOffer(secondOffer), this[1])
        }
    }
}